### PR TITLE
Fix PPL eval command string concatenation with + operator

### PR DESCRIPTION
**Pull request by ahkcs**
# Fix PPL eval command string concatenation with + operator

## Description

This PR fixes a critical issue where string concatenation using the + operator in PPL eval commands would fail with a runtime error during Calcite query planning. Related #4014

## Problem

PPL queries with string concatenation using the + operator were failing with:
```
java.sql.SQLException: Error while preparing plan
LogicalProject... full_name=[+(+($1, ' '), $10)]
```

Example failing query:
```sql
source=accounts | eval full_name = firstname + " " + lastname | fields account_number, firstname, lastname, full_name
```

## Example Usage

After this fix, the following PPL queries now work correctly:

```sql
# Basic string concatenation
source=accounts | eval full_name = firstname + " " + lastname
```

### CODE REVIEW

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java [hunk 1]
```diff
@@ -488,6 +488,22 @@ public RexNode visitFunction(Function node, CalcitePlanContext context) {
...
+    // Handle string concatenation with + operator
+    if ("+".equals(node.getFuncName()) && arguments.size() == 2) {
+      RexNode left = arguments.get(0);
+      RexNode right = arguments.get(1);
+
+      // Check if both operands are strings
+      if (left.getType().getSqlTypeName() == SqlTypeName.VARCHAR
+          || left.getType().getSqlTypeName() == SqlTypeName.CHAR) {
+        if (right.getType().getSqlTypeName() == SqlTypeName.VARCHAR
+            || right.getType().getSqlTypeName() == SqlTypeName.CHAR) {
```
Comment (ykmr1224) [inline core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: Do we convert types implicitly? In that case, should we handle other types?

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java [hunk 2]
```diff
@@ -488,6 +488,22 @@ public RexNode visitFunction(Function node, CalcitePlanContext context) {
       }
     }
 
+    // Handle string concatenation with + operator
+    if ("+".equals(node.getFuncName()) && arguments.size() == 2) {
+      RexNode left = arguments.get(0);
```
Comment (ykmr1224) [inline core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: Let's extract and make `visitFunction` smaller for better readability.

Comment (dai-chen) [reply core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: +1. Is it possible to leverage current visitor pattern to dispatch instead of adding more and more conditional checks in a single visit method?

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java (hunk 1 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: Currently, we don't perform implicit type conversion - we only handle the case where both operands are already string types (VARCHAR/CHAR) or both numbers(already supported)

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java (hunk 2 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: I have implemented a dispatch method to resolve this

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java [hunk 3]
```diff
@@ -497,6 +503,64 @@ public RexNode visitFunction(Function node, CalcitePlanContext context) {
...
+   */
+  private RexNode tryHandleSpecialOperators(
+      Function node, List<RexNode> arguments, CalcitePlanContext context) {
+    switch (node.getFuncName()) {
+      case "+":
+        return tryHandleStringConcatenation(arguments, context);
+      default:
+        return null;
+    }
+  }
```
Comment (LantaoJin) [inline core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: This fixing can address the request but the exception logic is tricky to me. It skips the TypeChecker. Can we refactor it similar to https://github.com/opensearch-project/sql/pull/3821? cc @yuancu for further suggestions

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java [hunk 4]
```diff
@@ -354,4 +354,24 @@ public void testDependedLateralEval() {
             + "GROUP BY `DEPTNO`";
     verifyPPLToSparkSQL(root, expectedSparkSql);
   }
+
+  @Test
+  public void testEvalStringConcatenationWithPlus() {
```
Comment (Swiddis) [inline ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java]: Do these tests cover what happens if there are null entries in the dataset?

Comment (ahkcs) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java]: I have added more tests to cover null entries scenario

File: core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java (hunk 3 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/calcite/CalciteRexNodeVisitor.java]: I have refactored

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java [hunk 5]
```diff
@@ -760,8 +760,36 @@ void populate() {
       registerOperator(AND, SqlStdOperatorTable.AND);
       registerOperator(OR, SqlStdOperatorTable.OR);
       registerOperator(NOT, SqlStdOperatorTable.NOT);
-      registerOperator(ADD, SqlStdOperatorTable.PLUS);
-      registerOperator(ADDFUNCTION, SqlStdOperatorTable.PLUS);
+
+      // Register ADD with custom implementation that handles both numeric and string types
+      register(
```
Comment (qianheng-aws) [inline core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java:694]: I would recommend registering  `+` for string in a separate function signature. The current builder should support registering multiple times for the same function name with different signatures. And then its type checker should be simpler, should using PPLFamilyTypeChecker directly.

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java [hunk 6]
```diff
@@ -760,8 +760,36 @@ void populate() {
...
+
+                if (arg1IsString && arg2IsString) {
+                  return builder.makeCall(SqlStdOperatorTable.CONCAT, arg1, arg2);
+                } else {
+                  return builder.makeCall(SqlStdOperatorTable.PLUS, arg1, arg2);
+                }
+              },
+              new PPLTypeChecker.PPLAddTypeChecker()));
+      register(
+          ADDFUNCTION,
```
Comment (qianheng-aws) [inline core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: Do we really want register `ADD` function for string?

**Review by qianheng-aws — CHANGES_REQUESTED** [review]

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java [hunk 7]
```diff
@@ -760,8 +760,36 @@ void populate() {
       registerOperator(AND, SqlStdOperatorTable.AND);
       registerOperator(OR, SqlStdOperatorTable.OR);
       registerOperator(NOT, SqlStdOperatorTable.NOT);
-      registerOperator(ADD, SqlStdOperatorTable.PLUS);
-      registerOperator(ADDFUNCTION, SqlStdOperatorTable.PLUS);
+
+      // Register ADD with custom implementation that handles both numeric and string types
+      register(
+          ADD,
+          createFunctionImpWithTypeChecker(
```
Comment (yuancu) [inline core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: This function is removed in #4024 as it is unnecessarily involved. Please refactor accordingly following #4024.

File: core/src/main/java/org/opensearch/sql/expression/function/PPLTypeChecker.java [hunk 8]
```diff
@@ -601,4 +601,71 @@ private static String formatExprSignatures(List<List<ExprType>> signatures) {
                     + "]")
         .collect(Collectors.joining(","));
   }
+
+  /**
+   * A custom type checker for the ADD operator that supports both numeric addition and string
+   * concatenation.
+   */
+  class PPLAddTypeChecker implements PPLTypeChecker {
```
Comment (yuancu) [inline core/src/main/java/org/opensearch/sql/expression/function/PPLTypeChecker.java]: Please remove this from `PPLTypeChecker` interface. Please use `PPLTypeChecker.wrapComposite` instead.

**Review by yuancu — CHANGES_REQUESTED** [review]
I suggest you re-implement this PR in the following way (after #4024 is approved):

```java
register(
        ADD,
        (RexBuilder builder, RexNode... args) ->
                builder.makeCall(SqlStdOperatorTable.PLUS, args),
        PPLTypeChecker.wrapFamily(OperandTypes.NUMERIC_NUMERIC));
register(
        ADD,
        (RexBuilder builder, RexNode... args) ->
                builder.makeCall(SqlStdOperatorTable.CONCAT, args),
        PPLTypeChecker.wrapFamily(OperandTypes.STRING_STRING));
```


This registers the same function with two different implementations, controlled via type checkers.


P.s. please make the PR title more specific

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java (hunk 7 above)
Comment (qianheng-aws) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: I think just `registerOperator(PLUS, SqlStdOperatorTable.CONCAT);` could implement this without specifying type checker, does it? @yuancu 

`registerOperator` already has capability to generate type checker itself automatically.

Comment (yuancu) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: You are right, I updated my suggestions on implementation to the following:

```java
register(
        ADD,
        (RexBuilder builder, RexNode... args) ->
                builder.makeCall(SqlStdOperatorTable.PLUS, args),
        PPLTypeChecker.wrapFamily(OperandTypes.NUMERIC_NUMERIC));

registerOperator(ADD, SqlStdOperatorTable.CONCAT);
```

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java (hunk 6 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: Removed

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java (hunk 7 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: Updated implementation

File: core/src/main/java/org/opensearch/sql/expression/function/PPLTypeChecker.java (hunk 8 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLTypeChecker.java]: Removed

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java (hunk 5 above)
Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java:694]: Updated implementation

File: core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java [hunk 9]
```diff
@@ -689,7 +689,21 @@ void populate() {
...
+          ADD,
+          (RexBuilder builder, RexNode... args) -> builder.makeCall(SqlStdOperatorTable.PLUS, args),
+          new PPLTypeChecker.PPLFamilyTypeChecker(SqlTypeFamily.NUMERIC, SqlTypeFamily.NUMERIC));
+
+      // Register ADD (+ symbol) for string concatenation
+      register(
+          ADD,
+          (RexBuilder builder, RexNode... args) ->
+              builder.makeCall(SqlStdOperatorTable.CONCAT, args),
+          new PPLTypeChecker.PPLFamilyTypeChecker(SqlTypeFamily.STRING, SqlTypeFamily.STRING));
```
Comment (yuancu) [inline core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: Is there any specific reason that you did not leverage `registerOperator(ADD, SqlStdOperatorTable.CONCAT);`?

> P.s. please make the PR title more specific. I.e. what enhancement it is.

A reminder ⬆️

P.s. please take some rest on Sunday 😂

Comment (ahkcs) [reply core/src/main/java/org/opensearch/sql/expression/function/PPLFuncImpTable.java]: I assumed both registrations needed explicit type checking, I have made changes according to your suggestion

P.S. Thanks for the reminder about rest! 😄 Will update the PR title to be more specific.

**Review by Swiddis — DISMISSED** [review]

**Review by ykmr1224 — DISMISSED** [review]

**Review by yuancu — COMMENTED** [review]
I think there should be an integration test to verify result correctness of string concatenation with +.

**Review by LantaoJin — DISMISSED** [review]

**Comment by LantaoJin** [comment]
> I think there should be an integration test to verify result correctness of string concatenation with +.

@ahkcs normally, we don't use `CalcitePPL*Test` to verify the result since the `CalcitePPL*Test` leverages Calcite catalog instead of ours which may have some logic gaps. `CalcitePPL*Test` is useful to verify the pure Calcite plan and SparkSQL generated. Instead, you can add an integration test in `CalcitePPL*IT`

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java (hunk 4 above)
Comment (yuancu) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java]: I doubt that these tests really cover cases where null entries exist -- these tests only verify the correctness of plans. It's only during the execution of these plans that one will know what happens when there are null operands.

**Comment by ahkcs** [comment]
@qianheng-aws @yuancu Added IT tests

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java [hunk 10]
```diff
@@ -354,4 +354,102 @@ public void testDependedLateralEval() {
...
+
+    String expectedSparkSql =
+        "SELECT `EMPNO`, `ENAME`, `JOB`, `ENAME` || ' ' || `JOB` `full_name`\n"
+            + "FROM `scott`.`EMP`\n"
+            + "LIMIT 3";
+    verifyPPLToSparkSQL(root, expectedSparkSql);
+  }
+
+  @Test
+  public void testEvalStringConcatenationWithNullableField() {
```
Comment (yuancu) [inline ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java]: Can be removed since concatenating string with null operands is actually verified in IT

**Review by LantaoJin — DISMISSED** [review]

File: ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java (hunk 10 above)
Comment (ahkcs) [reply ppl/src/test/java/org/opensearch/sql/ppl/calcite/CalcitePPLEvalTest.java]: Removed

**Review by qianheng-aws — APPROVED** [review]

**Review by LantaoJin — APPROVED** [review]
